### PR TITLE
Make sure there is no ongoing pod deletion before downscaling

### PR DESCRIPTION
**Pull request by sebgl**
We had a race condition were we would run downscales based on wrong
assumptions, if StatefulSets replicas we work with do not match pods
that are still alive from a previous downscale.

When we update the Replicas of a StatefulSet, it may take a while before
the corresponding pod is actually deleted. Since most of the downscale
logic relies on manipulating StatefulSets, this gives us wrong
assumpations to work with.

For example, we may end up:
- clearing shards allocation excludes, when a node from which data
was migrated away is not removed from the cluster yet, effectively
allowing it to allocate shards again
- updating zen1 minimum_master_nodes with a wrong value if a previous
master node deletion is not over yet
- considering we are in the 2->1 master nodes zen1 situation, but we are
actually in 3->1 since one master isn't removed yet
- removing 2 masters nodes at once

We could work with Pods instead of StatefulSets, but the logic would be a bit more complex
Instead, let's just requeue the downscale until pods we expect from StatefulSets spec are there.

To do that, we just compare the list of pods we have with the list of
pods we would expect according to StatefulSets we work with. If there is
a mismatch, we requeue. Seems simpler than any event expectations mechanism.

This also accounts for pods creation mismatch (a pod that was requested
for creation is not there yet), which, I believe, are not that much of a
problem for downscales purpose; but I think it's better to err on the
safe side here (do nothing if creations are not complete).

Fixes https://github.com/elastic/cloud-on-k8s/issues/1523.

### CODE REVIEW

File: operators/pkg/controller/elasticsearch/sset/list.go [hunk 1]
```diff
@@ -68,11 +70,11 @@ func (l StatefulSetList) PodNames() []string {
 	return names
 }
 
-// GetActualPods returns the list of pods currently existing in the StatefulSetList.
+// GetActualPodsForStatefulSet returns the list of pods currently existing in the StatefulSetList.
```
Comment (thbkrkr) [inline operators/pkg/controller/elasticsearch/sset/list.go]: Unsync godoc.

File: operators/pkg/controller/elasticsearch/driver/downscale.go [hunk 2]
```diff
@@ -67,6 +75,19 @@ func HandleDownscale(
...
+	// Pods we have may not match replicas specified in the StatefulSets spec.
+	// This can happen if, for example, replicas were recently downscaled to remove a node,
+	// but the node isn't completely terminated yet, and may still be part of the cluster.
+	// Moving on with downscaling more nodes may lead to complications when dealing with
+	// Elasticsearch shards allocation excludes (should not be cleared if the ghost node isn't removed yet)
+	// or zen settings (must consider terminating masters that are still there).
+	// Let's retry once expected pods are there.
+	// MatchActualPods also matches any pod not created yet, for which we'll also requeue.
+	return actualStatefulSets.MatchActualPods(downscaleCtx.k8sClient, downscaleCtx.es)
+}
```
Comment (thbkrkr) [inline operators/pkg/controller/elasticsearch/driver/downscale.go]: I'm in favor of small functions but here I wonder if it's worth it to have `noOnGoingDeletion` and `canClearVotingConfigExclusions` that just call `actualStatefulSets.MatchActualPods` and bring a little additionnal indirection to follow the flow of the execution?

File: operators/pkg/controller/elasticsearch/sset/list.go [hunk 3]
```diff
@@ -81,6 +83,22 @@ func (l StatefulSetList) GetActualPods(c k8s.Client) ([]corev1.Pod, error) {
 	return allPods, nil
 }
 
+// MatchActualPods returns true if actual existing pods match what is specified in the StatefulSetList.
+// It may return false if there are pods in the process of being created (but not created yet)
+// or terminated (but not removed yet).
+func (l StatefulSetList) MatchActualPods(c k8s.Client, es v1alpha1.Elasticsearch) (bool, error) {
```
Comment (thbkrkr) [inline operators/pkg/controller/elasticsearch/sset/list.go]: Isn't it more `MatchExpectedPods`?
Trying to find another name: `ArePodsReconciled`, `IsPodsReconcilationDone` or just `IsReconciled`.

File: operators/pkg/controller/elasticsearch/driver/downscale.go (hunk 2 above)
Comment (sebgl) [reply operators/pkg/controller/elasticsearch/driver/downscale.go]: My reasoning was that the big part of this function is actually the comment block that comes along with it. Including it in `HandleDownscale` kind of breaks the flow of reading `HandleDownscale`, which I see as the high-level flow of downscales.
I think `HandleDownscale` is nice to read as just "is there any ongoing deletion? No? ok continue". And if you want the details you get into the function implementation, read the comment to understand it's tricky, and realize it's just a call to `MatchActualPods` so we're lucky it's that simple.
WDYT?

File: operators/pkg/controller/elasticsearch/sset/list.go (hunk 3 above)
Comment (sebgl) [reply operators/pkg/controller/elasticsearch/sset/list.go]: I agree, that's confusing 👍 
I think I like `IsPodsReconcilationDone`, maybe simplified slightly to `.PodReconciliationDone()` (singular form).

File: operators/pkg/controller/elasticsearch/sset/list.go [hunk 4]
```diff
@@ -81,6 +83,22 @@ func (l StatefulSetList) GetActualPods(c k8s.Client) ([]corev1.Pod, error) {
...
+	// pods we expect to be there based on StatefulSets spec
+	expectedPods := l.PodNames()
+	// pods that are there for this cluster
+	actualRawPods, err := GetActualPodsForCluster(c, es)
+	if err != nil {
+		return false, err
+	}
+	actualPods := k8s.PodNames(actualRawPods)
+	// check if they match
+	return len(expectedPods) == len(actualPods) && stringsutil.StringsInSlice(expectedPods, actualPods), nil
```
Comment (pebrc) [inline operators/pkg/controller/elasticsearch/sset/list.go]: Would it make sense to have a debug log statement to find out which pods are keeping us from making progress?

**Pull request merged**